### PR TITLE
EDGECLOUD-4940 matchingenginehelper lib, EdgeEvents, Driving Mode

### PR DESCRIPTION
**Pull request by brucearmstrong**
* New matchinenginehelper library which encapsulates all MatchingEngine actions. This was to avoid duplicating a bunch of code when writing the TritonClient app. So now, the same matchingenginehelper code is used by all of:
  * The cloudlet map activity
  * The computervision library
  * The TritonClient app
* The matchinenginehelper has its own string values and preferences. EventLogViewer is also moved here, and accessed by the other activities/apps.
* Added EdgeEvents code to matchinenginehelper. When FindCloudletEvent is received, new cloudlet is marked on the map, and a connection test is performed. All EdgeEvents received are added to the EventLogViewer.
* Added ConnectionTester class for testing connection to app instances.
* New "Route Mode" menu items where you can choose either Driving or Flying. When one is selected,
  * Starting and Ending markers are placed on the map, with a route drawn between them.
  * A Play/Pause button is added to the UI to start and stop route animation.
  * If FindCloudlet has been performed, EdgeEvents are subscribed to, and the new location along the route is sent periodically.
  * If there are multiple cloudlets on the map, you can draw a route that will switch between which is closest, for an animated, visual demo.

### CODE REVIEW

File: android/MobiledgeXSDKDemo/app/src/main/java/com/mobiledgex/sdkdemo/qoe/QoeMapActivity.java [hunk 1]
```diff
@@ -1078,4 +1086,40 @@ public ColoredPoint(LatLng coords, float dlSpeed, float upSpeed) {
...
+
+    @Override
+    public void showError(String text) {
+
+    }
+
+    @Override
+    public void getCloudlets(boolean clearExisting) {
+
+    }
```
Comment (ah1053) [inline android/MobiledgeXSDKDemo/app/src/main/java/com/mobiledgex/sdkdemo/qoe/QoeMapActivity.java:1123]: Do you still need this ?

Comment (brucearmstrong) [reply android/MobiledgeXSDKDemo/app/src/main/java/com/mobiledgex/sdkdemo/qoe/QoeMapActivity.java:1123]: Oh, yeah. I need to rework that interface. This particular class only uses the MatchingEngine for getQosPositionKpi(). Nothing else.

File: android/MobiledgeXSDKDemo/matchingenginehelper/src/androidTest/java/com/mobiledgex/matchingenginehelper/ExampleInstrumentedTest.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ */
+@RunWith(AndroidJUnit4.class)
+public class ExampleInstrumentedTest {
+    @Test
+    public void useAppContext() {
+        // Context of the app under test.
+        Context appContext = InstrumentationRegistry.getInstrumentation().getTargetContext();
+        assertEquals("com.mobiledgex.matchingenginehelper.test", appContext.getPackageName());
+    }
+}
```
Comment (ah1053) [inline android/MobiledgeXSDKDemo/matchingenginehelper/src/androidTest/java/com/mobiledgex/matchingenginehelper/ExampleInstrumentedTest.java:26]: nit: add line at EOF

File: android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java [hunk 3]
```diff
@@ -0,0 +1,1053 @@
...
+
+        // Set a default Edge Events config.
+        // There is also a parameterized version to further customize.
+        EdgeEventsConfig backgroundEdgeEventsConfig = me.createDefaultEdgeEventsConfig();
+        backgroundEdgeEventsConfig.latencyTestType = NetTest.TestType.CONNECT;
+        // This is the internal port, that has not been remapped to a public port for a particular appInst.
+        backgroundEdgeEventsConfig.latencyInternalPort = DEFAULT_LATENCY_PORT;
+        backgroundEdgeEventsConfig.latencyUpdateConfig.maxNumberOfUpdates = 10; // Or Long.MAX_VALUE if you want. Default is 0.
+        backgroundEdgeEventsConfig.latencyUpdateConfig.updateIntervalSeconds = 7; // The default is 30.
+//        backgroundEdgeEventsConfig.latencyThresholdTrigger = 300;
```
Comment (ah1053) [inline android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java:186]: Why this is commented out ?

File: android/MobiledgeXSDKDemo/matchingenginehelper/src/test/java/com/mobiledgex/matchingenginehelper/ExampleUnitTest.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+ * Example local unit test, which will execute on the development machine (host).
+ *
+ * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
+ */
+public class ExampleUnitTest {
+    @Test
+    public void addition_isCorrect() {
+        assertEquals(4, 2 + 2);
+    }
+}
```
Comment (ah1053) [inline android/MobiledgeXSDKDemo/matchingenginehelper/src/test/java/com/mobiledgex/matchingenginehelper/ExampleUnitTest.java:17]: nit: add line at EOF

**Review by ah1053 — APPROVED** [review]
👍

File: android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java (hunk 3 above)
Comment (brucearmstrong) [reply android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java:186]: It was something Leon suggested I add, during debugging. Didn't make a difference for the case I was testing. Will remove or uncomment.

File: android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java [hunk 5]
```diff
@@ -0,0 +1,1060 @@
...
+            meHelperInterface.showMessage(message);
+        }
+
+        void handleFindCloudletServerPush(FindCloudletEvent event) {
+            meHelperInterface.showMessage("Received: FindCloudletEvent");
+
+            // In this demo case, use our existing interface to display the newly selected cloudlet on the map.
+            mClosestCloudlet = event.newCloudlet;
+            onFindCloudlet(mClosestCloudlet);
+        }
```
Comment (franklin-huang-mobiledgex) [inline android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java:874]: hmm. @lgarner Does he need this? Shouldn't the server push FindCloudletEvent be sent to the onMessageEvent?

**Review by franklin-huang-mobiledgex — APPROVED** [review]
This is awesome!

File: android/MobiledgeXSDKDemo/app/src/main/java/com/mobiledgex/sdkdemo/MainActivity.java [hunk 6]
```diff
@@ -299,35 +304,153 @@ public void onClick(View view) {
...
+            }
+            mRouteIsPlaying = false;
+        } else {
+            fabPlayRoute.setImageResource(R.drawable.ic_baseline_pause_24);
+            mEventLogViewer.mAutoExpand = false;
+            mUserLocationMarker.setPosition(mStartLatLng);
+            if (mRouteMode == RouteMode.FLYING) {
+                animateMarker(mRouteMode, mUserLocationMarker, mEndLatLng, 4000);
+            } else if (mRouteMode == RouteMode.DRIVING) {
+                animateMarker(mRouteMode, mUserLocationMarker, mEndLatLng, 10000);
```
[the remark this answers is not in the document]
Comment (brucearmstrong) [reply android/MobiledgeXSDKDemo/app/src/main/java/com/mobiledgex/sdkdemo/MainActivity.java:345]: Yes. Will also be adding preferences for these values.

File: android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java (hunk 5 above)
Comment (brucearmstrong) [reply android/MobiledgeXSDKDemo/matchingenginehelper/src/main/java/com/mobiledgex/matchingenginehelper/MatchingEngineHelper.java:874]: @franklin-huang-mobiledgex, Though originally written for the "give me all events" mode, this is also called from onMessageEvent(), so yes, it's still needed.

**Pull request merged**